:memo: updated getting started / secrets and variables (#6)
writechoiceorg/terrateam: docs/src/content/docs/getting-started/secrets-and-variables.mdx

title: Secrets and Variables
description: Using secrets, variables, and tfvars in Terrateam
---

> import { Tabs, TabItem } from '@astrojs/starlight/components';

Reviewer: ```suggestion
import { Tabs, TabItem } from '@astrojs/starlight/components';
import { Aside } from '@astrojs/starlight/components';
```

> Terrateam provides several ways to manage sensitive information and customize your Terraform configurations using secrets, variables, and tfvars files. This guide will walk you through how to create and use secrets, set environment variables, and use tfvars files to manage your Terraform variables.

Reviewer: ```suggestion
Terrateam offers multiple ways to manage sensitive information and customize your Terraform configurations using GitHub secrets, variables, or `.tfvars` files. All methods produce the same outcome, allowing you to choose the approach that best fits your workflow and needs for managing Terraform variables.
```

## GitHub Secrets and Variables

### Environment Variables

GitHub [Secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets) and [Variables](https://docs.github.com/en/actions/learn-github-actions/variables) can be used to set environment variables. They are translated into environment variables in the Terrateam GitHub Action runtime environment. These environment variables may be referenced in your Terraform code.

> **Example: Defining a GitHub Secret**

Reviewer: Transform the following code block into a guide. It does not make any sense to present it as a code block.

Reviewer: Here's an example you can use (I didn't use it as a suggestion because since it has a code block, it would break the suggestion):

You can use the GitHub CLI or the web interface to define a GitHub secret. 

The following code snippet shows the command you can use to generate a `AWS_ACCESS_KEY_ID` secret using GitHub CLI:

```
gh secret set AWS_ACCESS_KEY_ID --body "AKIAIOSFODNN7EXAMPLE"
```

If you want to use the web interface, follow the steps:
1. Access the repository with the Terrateam app installed.
2. Navigate to **Settings → Secrets and variables → Actions**.
3. Click on **New repository secret**.
4. Specify the **Name** (`AWS_ACCESS_KEY_ID`) and a **Secret** value (`AKIAIOSFODNN7EXAMPLE`). 
5. Click **Add secret**.

![image](https://github.com/user-attachments/assets/0674ab87-153f-4571-bb9e-ae02d4b5a79a)

```bash
# Using GitHub CLI to create a secret
gh secret set AWS_ACCESS_KEY_ID --body "AKIAIOSFODNN7EXAMPLE"

# Using GitHub Web Interface
# Navigate to your repository → Settings → Secrets and variables → Actions → New repository secret
# Name: AWS_ACCESS_KEY_ID
# Value: AKIAIOSFODNN7EXAMPLE
```

> **Example: Referencing the secret in Terraform**

Reviewer: ```suggestion
After adding a secret, you can reference it in Terrateam. The following code snippet shows how to use the previously created secret (`AWS_ACCESS_KEY_ID`) in the [ADD THE FILE NAME HERE].
```

Reviewer: @lucaspr2 , precisa atualizar o nome ali na frase.

```hcl
provider "aws" {
  region     = "us-west-2"
  access_key = var.aws_access_key_id
  secret_key = var.aws_secret_access_key
}

variable "aws_access_key_id" {
  description = "AWS Access Key ID"
  type        = string
  sensitive   = true
}

variable "aws_secret_access_key" {
  description = "AWS Secret Access Key"
  type        = string
  sensitive   = true
}
```

### Using TF_VAR Environment Variables

*Secrets and variables that start with* `TF_VAR_` *are treated specially by Terrateam.*

GitHub secrets and variables are always uppercase, however by convention Terraform variables are lowercase. Terrateam automatically finds all secrets that start with `TF_VAR_` and creates a new environment variable that has the lowercase name. If the lowercase name exists, no action is taken. The uppercase environment variable is left unchanged. 

> **Example:**

Reviewer: Avoid using this approach. Add a sentence.

Reviewer: ```suggestion
If you use the GitHub CLI to generate a secret named `TF_VAR_DATABASE_PASSWORD`:
```

```bash
# Setting a GitHub secret
gh secret set TF_VAR_DATABASE_PASSWORD --body "super-secret-password"
```

> This will create an environment variable `TF_VAR_database_password` which Terraform can automatically map to a variable named `database_password`.

Reviewer: ```suggestion
This will create an environment variable `TF_VAR_database_password` which Terraform can automatically map to a variable named `database_password` in your Terraform configuration.
```

```hcl

> # In your Terraform configuration

Reviewer: ```suggestion
```

variable "database_password" {
  description = "Password for database connection"
  type        = string
  sensitive   = true
}

resource "aws_db_instance" "example" {
  # Other configuration...
  password = var.database_password
}
```

### Security

GitHub secrets are stored encrypted at rest using a [libsodium sealed box](https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes) and are only decrypted when used in GitHub Actions workflows. 

> When Terrateam runs your Terraform operations:

Reviewer: I liked the approach. However, since not all elements are part of the flow.

> 1. Encrypted secrets are decrypted within the GitHub Actions runtime

Reviewer: ```suggestion
1. Encrypted secrets are decrypted within the GitHub Actions runtime.
```

> 2. Decrypted secrets are available as environment variables within the GitHub Actions runner

Reviewer: ```suggestion
2. Decrypted secrets are available as environment variables within the GitHub Actions runner.
```

> 3. Your sensitive data appears only in the encrypted Terraform plan files

Reviewer: ```suggestion
```

> 4. Terrateam does not store your unencrypted secrets on its servers

Reviewer: ```suggestion
<Aside type="tip" title="Your data is safe!">
Your sensitive data appears only in the encrypted Terraform plan files.
Terrateam does not store your unencrypted secrets on its servers.
</Aside>
```

        name: BAZ
        cmd: ['echo', 'QUX']
```

> **Example use case:** Setting a dynamic environment variable based on the current Git branch:

Reviewer: ```suggestion
The following code snippet shows how to set a dynamic environment variable based on the current Git branch:
```

      - type: apply
```

> **Example use case:** Setting a timestamp for resource tagging:

Reviewer: ```suggestion
The following code snippet shows how to set a timestamp for resource tagging:
```

```yaml
workflows:
  - tag_query: ""
    plan:
      - type: init
      - type: env
        name: TF_VAR_deployment_timestamp
        cmd: ["date", "+%Y%m%d%H%M%S"]
      - type: plan
```

## `.tfvars` Files

Terraform allows you to define variables in `.tfvars` files, which can be used to customize your Terraform configurations.

> To use a tfvars file with Terrateam, you can specify the file path in your workflow configuration using the `extra_args` option.

Reviewer: ```suggestion
To use a `.tfvars` file with Terrateam, you can specify the file path in your workflow configuration using the `extra_args` option.
```

      - type: init
      - type: apply
```

> **Example of a `.tfvars` file:**

Reviewer: ```suggestion
The following code snippet shows an example of a `.tfvars` file:
```
